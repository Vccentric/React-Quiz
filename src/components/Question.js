import React from "react";
import PropTypes from "prop-types";
import AnswerChoice from "./AnswerChoice";

// question component
const Question = ({
  currentQuestion,
  totalQuestions,
  quizStats,
  setQuizStats,
  setQuizList,
}) => {
  // setup variables
  const { currentIndex, totalCorrectAnswers, selectedAnswer } = quizStats;
  const { question, answerChoices, correctAnswer } = currentQuestion;
  const isDisabled = selectedAnswer === null;

  // setup functions
  const next = () => {
    const nextIndex = currentIndex + 1;
    const count =
      selectedAnswer === correctAnswer
        ? totalCorrectAnswers + 1
        : totalCorrectAnswers;
    if (nextIndex < totalQuestions) {
      setQuizStats({
        ...quizStats,
        currentIndex: nextIndex,
        selectedAnswer: null,
        totalCorrectAnswers: count,
      });
    } else {
      setQuizStats({
        ...quizStats,
        currentIndex: 0,
        selectedAnswer: null,
        showResults: true,
        totalCorrectAnswers: count,
      });
    }
  };

  // setup elements
  const listAnswerChoices = answerChoices?.map((item, index) => {
    // setup variable
    const key = index * Math.random();
    const value = item || "";
    const selected = selectedAnswer === value;

    // setup functions
    const selectAnswer = () => {
      setQuizStats({
        ...quizStats,
        selectedAnswer: value,
      });
    };

    return (
      <li key={key}>
        <AnswerChoice
          value={value}
          selected={selected}
          selectAnswer={selectAnswer}
        />
      </li>
    );
  });

  return (
    <div className="question">
      <h2 className="index-total">
        Question {currentIndex + 1}/{totalQuestions}
      </h2>
      <p className="question-text">{question}</p>
      <ul>{listAnswerChoices}</ul>
      <div className="actions-container">
        <button className="action quit" onClick={setQuizList}>
          Quit
        </button>
        <button className="action" onClick={next} disabled={isDisabled}>
          Next
        </button>
      </div>
    </div>
  );
};

Question.propTypes = {
  currentQuestion: PropTypes.object,
  totalQuestions: PropTypes.number,
  quizStats: PropTypes.object,
  setQuizStats: PropTypes.func,
  setQuizList: PropTypes.func,
};

export default Question;
