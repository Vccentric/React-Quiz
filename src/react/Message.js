import React from 'react';
import { hot } from 'react-hot-loader';

// message component
function Message(props) {
    const { isAnswered, isCorrect } = props;
    return (
        <div
            id="message"
            style={isAnswered ? { display: 'block' } : { display: 'none' }}
            className={isCorrect ? 'correct' : 'wrong'}
        >
            {isCorrect ? 'Correct Answer!' : 'Wrong Answer!'}
        </div>
    );
}

export default hot(module)(Message);
