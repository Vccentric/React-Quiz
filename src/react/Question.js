import React from 'react';
import { hot } from 'react-hot-loader';

// question component
function Question(props) {
    const { question } = props;
    return <p id="question">{question}</p>;
}

export default hot(module)(Question);
