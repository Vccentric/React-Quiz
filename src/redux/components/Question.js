import React from 'react';
import { hot } from 'react-hot-loader';

// question component
const Question = ({ question }) => (
    <p id="question">{question}</p>
);

export default hot(module)(Question);
