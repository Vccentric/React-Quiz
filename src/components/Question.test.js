import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import React from "react";
import Question from "./Question";

// setup variables
const currentQuestion = {};
const totalQuestions = 5;
const quizStats = {
  currentIndex: 0,
  selectedAnswer: null,
  showResults: false,
  totalCorrectAnswers: 0,
};
const setQuizStats = jest.fn();
const setQuizList = jest.fn();

test("renders correctly - snapshot", () => {
  const tree = renderer
    .create(
      <Question
        currentQuestion={currentQuestion}
        totalQuestions={totalQuestions}
        quizStats={quizStats}
        setQuizStats={setQuizStats}
        setQuizList={setQuizList}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders results with mock values", () => {
  const mockQuestion = {
    question: "Question Name",
    answerChoices: ["100", "200"],
    correctAnswer: "100",
  };
  render(
    <Question
      currentQuestion={mockQuestion}
      totalQuestions={totalQuestions}
      quizStats={quizStats}
      setQuizStats={setQuizStats}
      setQuizList={setQuizList}
    />
  );
  expect(screen.getByText("Question Name")).toBeInTheDocument();
  expect(screen.getByText("100")).toBeInTheDocument();
});
