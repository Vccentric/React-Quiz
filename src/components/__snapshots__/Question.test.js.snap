// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly - snapshot 1`] = `
<div
  className="question"
>
  <h2
    className="index-total"
  >
    Question 
    1
    /
    5
  </h2>
  <p
    className="question-text"
  />
  <ul />
  <div
    className="actions-container"
  >
    <button
      className="action quit"
      onClick={[MockFunction]}
    >
      Quit
    </button>
    <button
      className="action"
      disabled={true}
      onClick={[Function]}
    >
      Next
    </button>
  </div>
</div>
`;
